atlasdb-commons/src/main/java/com/palantir/util/result/Result.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.util.result;
+
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+
+public abstract class Result<T, E> {
+
+    private Result() {}
+
+    public static final class Ok<T, E> extends Result<T, E> {
+        private final T value;
+
+        public Ok(T value) {
+            this.value = value;
+        }
+
+        public T getValue() {
+            return value;
+        }
+    }
+
+    public static final class Err<T, E> extends Result<T, E> {
+        private final E error;
+
+        public Err(E error) {
+            this.error = error;
+        }
+
+        public E getError() {
+            return error;
+        }
+    }
+
+    public boolean isOk() {
+        return this instanceof Ok;
```

Comment:
is there a reason all of these methods do instanceof and branching instead of just implementing the methods in each class?

Reply:
Regarding instanceOf: I didn't want more subclasses to potentially exist. So only `Ok` and `Err` should exist and that behaviour is enforced by the parent class checking instanceOf. Also to make it impossible to get wrong in different implementations overriding the `isOk` or `isErr` in ways that didn't make sense.

The branching was not really intentional and probably better refactored to be implemented by the subclasses.